Introducing EnvInjector!

alxhub opened the pull request:
This PR includes a couple commits which introduce the new `EnvInjector` concept. `EnvInjector` is a generalized version of `NgModuleRef` which represents the "module injector" (now renamed to "environment injector"). It's one of the primitives needed to support upcoming standalone components APIs.

AndrewKushnir committed:
feat(core): implement `importProvidersFrom` function

This commit implements the `importProvidersFrom` function that allows
extracting a list of `Provider`s from a list of NgModule types. The
R3Injector which implements DI at the "module" level for Angular is
refactored to use this functionality under the hood.

This commit also implements `INJECTOR_INITIALIZER`, a DI multi-provider
token which is used to run initialization logic when an injector is created.

alxhub committed:
feat(core): implement EnvironmentInjector with adapter to NgModuleRef

This commit exposes a new `EnvironmentInjector` abstraction, which
generalizes the "module injector" concept to injectors that are not based on
NgModules.

An EnvironmentInjector is a conceptual analogue of an `NgModuleRef` - it
represents an injector on the former "module" DI hierarchy in Angular (now
renamed to the "environment injector hierarchy"). Environment injectors are
created via the `createEnvironmentInjector` function from a list of
`Provider`s.

For backwards compatibility with current code using `NgModuleRef`,
`EnvironmentInjector`s are wrapped by an adapter `NgModuleRef`
implementation, so injecting `NgModuleRef` always returns the latest
`EnvironmentInjector`, even if that injector was not based on an NgModule.
Conversely, NgModule-based `NgModuleRef`s created via `createNgModuleRef`
are _also_ `EnvironmentInjector`s.

alxhub committed:
refactor(core): internally support `providedIn: environment`

This commit adds a new internal scope to `R3Injector` for
`EnvironmentInjector`s specifically. This will allow us to scope services to
the environment side of the injector hierarchy specifically, as opposed to
the `'any'` scope which also includes view-side injectors created via
`Injector.create`. For now, this functionality is not exposed publicly, but
is available to use within `@angular/core` only.

pkozlowski-opensource commented on packages/core/src/render3/interfaces/definition.ts:
```diff
@@ -7,6 +7,7 @@
  */
 
 import {ProcessProvidersFunction} from '../../di/interface/provider';
+import {EnvInjector} from '../../di/r3_injector';
```
nit: import without usage?

alxhub committed:
test(core): fix a crosstalk issue with locale ids

Previously there was a test ordering issue with the application_module_spec
tests where the value of `getLocaleId()` depended on the order in which
tests ran. Specifically, `setLocaleId()` lower-cases the current locale ID,
so the measured value in a test depended on whether a previous test had
called `setLocaleId()` (the difference between 'en-US' and 'en-us').

AndrewKushnir commented on goldens/public-api/core/index.md:
```diff
@@ -438,6 +441,17 @@ export abstract class EmbeddedViewRef<C> extends ViewRef {
+    // @deprecated (undocumented)
+    abstract get(token: any, notFoundValue?: any): any;
+    // (undocumented)
+    abstract onDestroy(callback: () => void): void;
```
I think we should provide an ability to unsubscribe, similar to the code here:
https://github.com/angular/angular/pull/45624/files#diff-c3bc2f4d2923398d46897c1233ab19cec2611e44ea60592e7e0d5ed03747a897R937

Not a blocker, we can do that in a followup PR.

alxhub replied:
Actually I marked this method `@internal` - we shouldn't make it public API (yet).

AndrewKushnir commented on goldens/public-api/core/index.md:
```diff
@@ -438,6 +441,17 @@ export abstract class EmbeddedViewRef<C> extends ViewRef {
+    // (undocumented)
+    abstract destroy(): void;
+    abstract get<T>(token: ProviderToken<T>, notFoundValue?: T, flags?: InjectFlags): T;
+    // @deprecated (undocumented)
+    abstract get(token: any, notFoundValue?: any): any;
```
Can we drop this deprecated signature from the `EnvironmentInjector` class?

alxhub replied:
I don't think so, because it's needed to implement the `Injector` interface.

alxhub replied:
I don't think so, because it's needed to implement the `Injector` interface.

AndrewKushnir commented on packages/compiler-cli/src/ngtsc/annotations/directive/src/handler.ts:
```diff
@@ -79,6 +79,9 @@ export class DirectiveDecoratorHandler implements
     // been processed, but we want to enforce a consistent decorator mental model.
     // See: https://v9.angular.io/guide/migration-undecorated-classes.
     if (this.compileUndecoratedClassesWithAngularFeatures === false && decorator === null) {
+      if (this.isCore) {
```
May be add a quick comment on why there is a special handling for core?

AndrewKushnir commented on packages/core/src/di/r3_injector.ts:
```diff
@@ -52,12 +50,20 @@ const NOT_YET = {};
 
+/**
+ * Special token that allows attaching a custom logic that would be invoked once an injector is
+ * created by the framework.
+ */
```
```suggestion
 * created by the framework.
 *
 * @publicApi
 */
```

AndrewKushnir commented on packages/core/src/di/r3_injector.ts:
```diff
@@ -95,35 +101,203 @@ export function createInjector(
+      for (let i = 0; i < importTypesWithProviders.length; i++) {
+        const {ngModule, providers} = importTypesWithProviders[i];
+        deepForEach(providers!, provider => {
+          validateProvider(provider, providers || EMPTY_ARRAY, ngModule);
```
Should we guard it with the `ngDevMode` flag?
```suggestion
          ngDevMode && validateProvider(provider, providers || EMPTY_ARRAY, ngModule);
```

AndrewKushnir commented on packages/core/src/di/r3_injector.ts:
```diff
@@ -95,35 +101,203 @@ export function createInjector(
+    const injectorType = container as InjectorType<any>;
+    deepForEach(defProviders, provider => {
+      // TODO: fix cast
+      validateProvider(provider, defProviders as any[], injectorType);
```
```suggestion
      ngDevMode && validateProvider(provider, defProviders as any[], injectorType);
```

alxhub replied:
I think this would be a behavior change - r3_injector previously validated providers even in prod mode.

AndrewKushnir commented on packages/core/src/di/r3_injector.ts:
```diff
@@ -95,35 +101,203 @@ export function createInjector(
+ * @publicApi
+ */
+export function importProvidersFrom(...injectorTypes: Array<Type<unknown>>): Provider[] {
+  const providers: SingleProvider[] = [];
```
```suggestion
  const providers: Provider[] = [];
```
(so that it's aligned with the return type)

alxhub replied:
`SingleProvider` doesn't allow recursion (nested provider arrays). It's a good distinction to have here.

AndrewKushnir commented on packages/core/src/di/r3_injector.ts:
```diff
@@ -95,35 +101,203 @@ export function createInjector(
+ */
+export function walkProviderTree(
+    container: InjectorType<unknown>|InjectorTypeWithProviders<unknown>,
+    providersOut: SingleProvider[], parents: InjectorType<unknown>[],
```
```suggestion
    providersOut: Provider[], parents: InjectorType<unknown>[],
```

alxhub replied:
Same as above.

AndrewKushnir reviewed (approved):
Looks great 👍 Just left a few minor comments (none of them are blockers, we can update things in followup PRs).

atscott commented on goldens/public-api/core/index.md:
```diff
@@ -1432,7 +1432,7 @@ export abstract class ViewRef extends ChangeDetectorRef {
 export function ɵɵdefineInjectable<T>(opts: {
     token: unknown;
-    providedIn?: Type<any> | 'root' | 'platform' | 'any' | null;
+    providedIn?: Type<any> | 'root' | 'platform' | 'any' | 'env' | null;
```
bike-shedding: I still think I would prefer `'environment'` here.

jessicajaniuk replied:
➕  1 to this

atscott reviewed (changes requested):
reviewed-for: public-api

atscott reviewed (approved):
reviewed-for: size-tracking

dylhunn reviewed (approved):
reviewed-for: public-api, size-tracking

jessicajaniuk commented on packages/core/src/di/interface/defs.ts:
```diff
@@ -36,7 +37,7 @@ export interface ɵɵInjectableDeclaration<T> {
    *   `providers`.
    */
-  providedIn: InjectorType<any>|'root'|'platform'|'any'|null;
+  providedIn: InjectorType<any>|'root'|'platform'|'any'|'env'|null;
```
Similar comment to earlier ones about this being `environment` instead of the abbreviation.

jessicajaniuk commented on packages/core/src/di/interface/defs.ts:
```diff
@@ -141,7 +142,7 @@ export interface InjectorTypeWithProviders<T> {
 export function ɵɵdefineInjectable<T>(opts: {
   token: unknown,
-  providedIn?: Type<any>|'root'|'platform'|'any'|null, factory: () => T,
+  providedIn?: Type<any>|'root'|'platform'|'any'|'env'|null, factory: () => T,
```
Here too.

jessicajaniuk commented on packages/core/src/di/scope.ts:
```diff
@@ -9,9 +9,11 @@
 import {InjectionToken} from './injection_token';
 
 
+export type InjectorScope = 'root'|'platform'|'env';
```
Here as well.

jessicajaniuk commented on packages/core/src/di/r3_injector.ts:
```diff
@@ -133,31 +307,34 @@ export class R3Injector {
     this.records.set(INJECTOR, makeRecord(undefined, this));
 
+    // And `EnvironmentInjector` if the current injector is supposed to be env-scoped.
+    if (scopes.has('env')) {
```
Here's another instance of `env`. Is there any way to make these not magic strings?

alxhub replied:
They're sort of by definition magic strings, since they're typically generated by the compiler.

jessicajaniuk commented on packages/core/src/di/r3_injector.ts:
```diff
@@ -95,35 +101,203 @@ export function createInjector(
+  const ngModule =
+      (def == null) && (container as InjectorTypeWithProviders<any>).ngModule || undefined;
+
+  // Determine the InjectorType. In the case where `defOrWrappedDef` is an `InjectorType`,
```
Just a loose comment to say I appreciate how well commented this is. Thank you for that!

jessicajaniuk commented on packages/core/src/render3/ng_module_ref.ts:
```diff
@@ -66,12 +68,12 @@ export class NgModuleRef<T> extends viewEngine_NgModuleRef<T> implements Interna
                              }
                            ],
-                           stringify(ngModuleType)) as R3Injector;
+                           stringify(ngModuleType), new Set(['env'])) as R3Injector;
```
The other locations would be env-ious if you didn't also change this one.

jessicajaniuk commented on packages/core/src/render3/ng_module_ref.ts:
```diff
@@ -105,3 +107,43 @@ export class NgModuleFactory<T> extends viewEngine_NgModuleFactory<T> {
+          {provide: viewEngine_NgModuleRef, useValue: this},
+          {provide: viewEngine_ComponentFactoryResolver, useValue: this.componentFactoryResolver},
+        ],
+        parent || getNullInjector(), source, new Set(['env']));
```
Apologies for the `env` spam

alxhub replied:
You're always `env`-ited to review as meticulously as you'd like.

jessicajaniuk reviewed (commented):
LGTM :cookie:

Just a few non-blocking comments.

jessicajaniuk reviewed (approved):
reviewed-for: public-api, size-tracking

pkozlowski-opensource commented on packages/core/src/core.ts:
```diff
@@ -36,7 +36,7 @@ export * from './core_private_export';
 export {SecurityContext} from './sanitization/security';
 export {Sanitizer} from './sanitization/sanitizer';
-export {createNgModuleRef} from './render3/ng_module_ref';
+export {createNgModuleRef, createEnvironmentInjector} from './render3/ng_module_ref';
```
nit: import without usage?

pkozlowski-opensource commented on packages/core/src/di/index.ts:
```diff
@@ -18,6 +18,7 @@ export {ɵɵdefineInjectable, defineInjectable, ɵɵdefineInjector, InjectableTy
 export {forwardRef, resolveForwardRef, ForwardRefFn} from './forward_ref';
 export {Injectable, InjectableDecorator, InjectableProvider} from './injectable';
 export {Injector} from './injector';
+export {EnvironmentInjector, INJECTOR_INITIALIZER} from '../di/r3_injector';
```
nit: import without usage?

pkozlowski-opensource commented on packages/core/src/render3/index.ts:
```diff
@@ -138,7 +138,7 @@ export {CssSelectorList, ProjectionSlots} from './interfaces/projection';
   setClassMetadata,
 } from './metadata';
-export {NgModuleFactory, NgModuleRef} from './ng_module_ref';
+export {NgModuleFactory, NgModuleRef, createEnvironmentInjector} from './ng_module_ref';
```
nit: import without usage?

alxhub replied:
This is an `export` :)

pkozlowski-opensource replied:
OMG, tired brain :-)

atscott reviewed (approved):
reviewed-for: public-api

alxhub commented:
Caretaker: this has all required approvals

dylhunn commented:
This PR was merged into the repository by commit 3578e943841f0ca0f8496ed06c901a72a8356f9f.
